src/security/KeyData.cc

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright (C) 1996-2017 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#include "squid.h"
+#include "anyp/PortCfg.h"
+#include "fatal.h"
+#include "security/KeyData.h"
+#include "SquidConfig.h"
+#include "ssl/bio.h"
+
+/// verify that a private key and cert match
+static bool
+checkPrivateKey(Security::CertPointer &cert, Security::PrivateKeyPointer &pkey)
+{
+#if USE_OPENSSL
+    return X509_check_private_key(cert.get(), pkey.get());
+#elif USE_GNUTLS
+    return true; // TODO find GnuTLS equivalent check
```

Comment:
I thought so too until I read what that OpenSSL function actually does - which is purely to check that parameters like KeyUsage *if any* on the key are also on the cert .... there can apparently be additional ones on the cert not in the key, and the two cert/key items can be completely unrelated while still passing this check.

The GnuTLS API does most of this kind of check automatically when you add that cert+key pair to the context instead of being explicitly done by us. So I think we are safe enough leaving it to do that for now.

The check this call *seemed* to be doing (verify the cert and key were actually a matching pair) is something we probably should implement for nicer squid.conf error messages. But that is also missing for OpenSSL so out of scope in this PR. I have an XXX in the caller about doing something better than this function can provide right now to verify the config details.

Reply:
> that OpenSSL function [...] is purely to check that parameters like KeyUsage if any on the key are also on the cert

AFAICT, comparing parameters is not the _only_ check that X509_check_private_key() performs in OpenSSL v1.1.0f. The function also checks that the public key of the given certificate matches the given private key (via an EVP_PKEY_cmp() call). For example, for RSA keys, EVP_PKEY_cmp() uses rsa_pub_cmp() to compare the two "raw" key values:

```
static int rsa_pub_cmp(const EVP_PKEY *a, const EVP_PKEY *b)
{
    if (BN_cmp(b->pkey.rsa->n, a->pkey.rsa->n) != 0
        || BN_cmp(b->pkey.rsa->e, a->pkey.rsa->e) != 0)
        return 0;
    return 1;
}
```

Why is OpenSSL comparing public and private keys? According to EVP_PKEY_cmp() manual page, "Since OpenSSL private keys contain public key components too the function
EVP_PKEY_cmp() can also be used to determine if a private key matches
a public key".

My code interpretation seems to match X509_check_private_key() [documentation](https://www.openssl.org/docs/manmaster/man3/X509_check_private_key.html) which says that the function returns success "if the keys match each other". There is a separate function for comparing just the parameters.

Did I misinterpret what X509_check_private_key() does?

Reply:
The section "BUGS" in that documentation you referenced explicitly mentions that passing it two _public_ keys (x and k) generated with the same parameters will return success.

Reply:
With the new structure of KeyData.cc members I have inlined this check to the applicable USE_OPENSSL code block.